drain_instance.py

```diff
@@ -28,16 +29,24 @@ def lambda_handler(event, context):
     if msg['LifecycleTransition'] != 'autoscaling:EC2_INSTANCE_TERMINATING':
         return
 
-    ciId, status = ciFor(ec2Id)
-    if ciId == None:
-        return
-
-    if status != 'DRAINING':
-        ecs.update_container_instances_state(cluster=clusterName,containerInstances=[ciId],status='DRAINING')
-
-    tasks = ecs.list_tasks(cluster=clusterName, containerInstance=ciId)['taskArns']
-    if len(tasks) > 0:
-        time.sleep(5)
+    # wait for random time to avoid ThrottlingException of AWS API call
+    sec = random.uniform(0, 5)
+    time.sleep(sec)
+
+    try:
+        ciId, status = ciFor(ec2Id)
+        if ciId == None:
+            return
+        if status != 'DRAINING':
+            ecs.update_container_instances_state(cluster=clusterName,containerInstances=[ciId],status='DRAINING')
+
+        tasks = ecs.list_tasks(cluster=clusterName, containerInstance=ciId)['taskArns']
+        if len(tasks) > 0:
```

Comment:
@essa 
Not sure if I misunderstood the algorithm.
Here, we are checking if the container instance (EC2 instance) still have any tasks running.
If `len(tasks)>0`, there are still tasks running, so we wait for 5 seconds.
After the 5 seconds, shouldn't we again do `tasks = ecs.list_tasks(...)` and check `len(tasks)>0`?
Shouldn't we have a `while len(tasks) > 0 ...`?

Reply:
> After the 5 seconds, shouldn't we again do tasks = ecs.list_tasks(...) and check len(tasks)>0?

Yes, and we are doing exactly what you are saying from the beginning of Barcelona. But we do it by triggering a call for this Lambda Function.

```python
 session.client('sns').publish(TopicArn=topicArn, Message=json.dumps(msg), Subject='Invoking lambda again')
```

This for reducing the execution time of Lambda. Stopping tasks will take five to ten minutes sometimes.

Reply:
Understood.